Refactor sysctl to use generic reconciler

pippolo84 opened the pull request:
Refactor the sysctl package using the statedb table and the generic
reconciler. The sysctl specific reconciler operations are in charge of
syncing the desired state with the actual kernel sysctl parameters.

To be backward compatible with the old sysctl interface, the exported
methods are kept synchronous, so they'll wait for the reconciliation to
actually happen before returning to the caller.

Moreover, ReadSysctl and WriteSysctl offer a low-level "direct"
implementation to change sysctl parameters without relying on the
reconciler. They are meant to be used *ONLY* in binaries that are not
based on the "hive and cells" framework, IOW where a statedb and the
reconcilers are not available (e.g: cilium-health and cilium-cni).

joamaki commented on pkg/sysctl/sysctl.go:
```diff
@@ -6,177 +6,122 @@ package sysctl
+		Val:    "0",
+		Status: reconciler.StatusPending(),
+	})
 
```
Wait until `name` has been reconciled (`FirstWatch` with query on `name`, check Status.Kind == StatusKindDone, wait on watch and check again (or timeout)). Applies to all the write APIs

joamaki commented on pkg/sysctl/sysctl.go:
```diff
@@ -6,177 +6,122 @@ package sysctl
-	if err != nil {
-		return "", fmt.Errorf("Failed to read %s: %s", path, err)
+func (mgr *Manager) Read(name string) (string, error) {
+	txn := mgr.db.ReadTxn()
```
Read directly from proc, or perhaps read from settings first and if nothing there then from proc?

pippolo84 committed:
cilium-dbg: Add sysctl table to statedb output

Add sysctl table to cilium-dbg statedb output and update related
documentation.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/linux/sysctl/sysctl.go:
```diff
@@ -0,0 +1,208 @@
+	case <-watch:
+	}
+	if obj.Status.Kind == reconciler.StatusKindError {
+		return errors.New(obj.Status.Error)
```
Hmm this makes me wonder if `Status.Error` should be `error`. I think the reason why I went with `string` was to make the default JSON dumping look reasonable, but perhaps it'd be nicer to keep it as `error` and just have custom JSON marshalling for `Status`...

pippolo84 replied:
I think having an `error` + custom JSON marshalling would be better. I expect many uses of `Status.Error` as the reconciler is used more.

joamaki replied:
Yeah agreed. Using `error` also allows delaying the stringification of the error and potentially allowing for greater reuse of error values and potentially just storing a pointer to a shared error value in `Status.Error`. This might be something we want to do for massive tables. (Other option is to keep it as string and do string interning in StatusError constructor).

joamaki commented on pkg/datapath/linux/sysctl/sysctl.go:
```diff
@@ -0,0 +1,208 @@
+		return nil
+	}
+
+	t := time.NewTimer(time.Second)
```
This should loop waiting for the status to turn to `StatusKindDone` rather than just look at the first change.

joamaki commented on pkg/datapath/linux/sysctl/cell.go:
```diff
@@ -0,0 +1,43 @@
+
+var Cell = cell.Module(
+	"sysctl",
+	"Tweak sysctl parameters and reconcile them with the kernel",
```
This reads a bit weird. Maybe just "Manages sysctl settings"?

joamaki commented on pkg/datapath/linux/sysctl/cell.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+func (cfg Config) Flags(flags *pflag.FlagSet) {
+	flags.String("procfs", "/proc", "Root's proc filesystem path")
```
I know this is the original doc string, but it's pretty weird. I have no idea what "Root" is here. Perhaps "Path to the proc filesystem" or "Proc filesystem mount point" or "Path to the host's proc filesystem mount"?

joamaki commented on pkg/datapath/linux/sysctl/reconciliation.go:
```diff
@@ -0,0 +1,82 @@
+			return true, fmt.Errorf("failed to sysctl -w %s=%s: %s", s.Name, s.Val, err)
+		}
+
+		warn := "Failed to sysctl -w"
```
`sysctl -w` is misleading since that's not what this does. Perhaps we could include the full path instead and say failed to write sysctl setting to `/proc/...`.

joamaki replied:
Though I do wonder if we really need to log at all here and it'd be better to instead have logging of failing health be done by the module health system. It can there decide how to for example rate-limit logging of consistently failing modules and such. Fine to leave the logging in place though and have such refactoring done later.

joamaki commented on pkg/datapath/linux/sysctl/reconciliation.go:
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
```
I would just call this file `ops.go` or `sysctl_ops.go` or some such.

joamaki commented on pkg/datapath/linux/sysctl/reconciliation.go:
```diff
@@ -0,0 +1,82 @@
+	"github.com/cilium/cilium/pkg/statedb/reconciler"
+)
+
+func newReconcilerConfig(
```
I would move this to `cell.go` as it's pretty useful to see these params at a glance there.

joamaki commented on pkg/datapath/linux/sysctl/reconciliation.go:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+func newReconcilerOps(log logrus.FieldLogger, fs afero.Fs, cfg Config) reconciler.Operations[*tables.Sysctl] {
```
just `newOps` is specific enough within this package

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -60,6 +61,9 @@ var Cell = cell.Module(
 	monitorAgent.Cell,
 
+	// The sysctl reconciler to read and write kernel sysctl parameters.
+	sysctl.Cell,
```
Hmm, do we need a no-op `Sysctl` implementation? I guess it'd be great if we didn't need one, but I suspect we might... Though if you don't find any cases where we do then let's just run with that.

pippolo84 replied:
I think we need one. It should become mandatory as soon as I'll finish the PR plumbing the new `sysctl.Sysctl` into the dependant cells.

pippolo84 committed:
test/controlplane: Add fake sysctl

Add a fake implementation of Sysctl interface to be used in controlplane
testing.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
datapath: Fix usage of bandwidth manager type

Replace usages of bandwidth.Manager type with types.BandwidthManager to
fix agent hive building:

```
failed to build types.Datapath: missing dependencies for function \"github.com/cilium/cilium/pkg/datapath\".newDatapath
(/home/pippolo/go/src/github.com/cilium/cilium/pkg/datapath/cells.go:159): missing type: bandwidth.Manager"
```

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/datapath/linux/sysctl/sysctl.go:
```diff
@@ -0,0 +1,234 @@
+
+	var err error
+	for {
+		select {
```
The `FirstWatch` must be within the loop. After a watch channel closes you always need a new transaction (ReadTxn()) and a new query.

joamaki commented on pkg/datapath/linux/sysctl/sysctl.go:
```diff
@@ -0,0 +1,234 @@
+		return nil
+	}
+
+	t := time.NewTimer(time.Second)
```
Move the `time.Second` into a constant with a comment and mention for each of the methods that call this that the method blocks until reconciled or timeout defined by this constant is reached.

pippolo84 committed:
sysctl: Move sysctl package under datapath/linux

Sysctl should be part of the datapath module, so the commit moves the
package accordingly.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
datapath/sysctl: Use generic reconciler for sysctl params

Refactor the sysctl package using the statedb table and the generic
reconciler. The sysctl specific reconciler operations are in charge of
syncing the desired state with the actual kernel sysctl parameters.

To be backward compatible with the old sysctl interface, the exported
methods are kept synchronous, so they wait for the reconciliation to
actually happen before returning to the caller.

Moreover, ReadSysctl and WriteSysctl offer a low-level "direct"
implementation to change sysctl parameters without relying on the
reconciler. They are meant to be used *ONLY* in binaries that are not
based on the "hive and cells" framework, IOW where a statedb and the
reconcilers are not available (e.g: cilium-health and cilium-cni).

The sysctl package tests are adapted to the new implementation, and the
TestSysctl is written as a unit-style integration test to verify the
sysctl reconciler operations.

All the privileged tests that don't rely on a hive but require a working
implementation of the sysctl package are temporarily left broken. The
same goes for the controlplane tests that require a fake sysctl
implementation.
They will be fixed in the subsequent commits.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
datapath/sysctl: Add a sysctl implementation for tests

Add an alternative implementation of sysctl to be used only in
privileged tests or benchmarks where a hive is not available.

Fix all the privileged tests and benchmarks using the newly added
implementation.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on pkg/wireguard/agent/agent.go:
```diff
@@ -71,6 +71,7 @@ type Agent struct {
 	ipCache        *ipcache.IPCache
 	listenPort     int
 	privKey        wgtypes.Key
+	sysctl         sysctl.Sysctl
```
this is not set anywhere and thus crashes in CI with nil deref.

joamaki commented on pkg/datapath/linux/sysctl/sysctl.go:
```diff
@@ -0,0 +1,253 @@
+				return nil
+			}
+			if obj.Status.Kind == reconciler.StatusKindError {
+				if obj.IgnoreErr {
```
This seems unnecessary to me since `Update` already handles `IgnoreErr`

joamaki commented on pkg/datapath/linux/sysctl/ops.go:
```diff
@@ -0,0 +1,72 @@
+
+	val, err := ReadSysctl(ops.fs, path)
+	if err != nil {
+		return false, err
```
should we check `IgnoreErr` here?

joamaki commented on pkg/datapath/linux/sysctl/ops.go:
```diff
@@ -0,0 +1,72 @@
+		return false, err
+	}
+	if val == s.Val {
+		log.Debugf("sysctl value already updated %s %s", s.Name, s.Val)
```
This can be dropped

rolinh reviewed (approved):
Vendor changes lgtm
